bark_ml/model_loader/ModelLoader.cpp

```diff
@@ -0,0 +1,99 @@
+#include "libtensorflow/include/tensorflow/c/c_api.h"
+#include "ModelLoader.hpp"
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <vector>
+#include <numeric>
+#include <string.h>
+
+void NoOpDeallocator(void* data, size_t a, void* b) {}
+
+ModelLoader::ModelLoader(const char* saved_model_dir)
+    {
+        //********* Read model
+        Graph = TF_NewGraph();
+        Status = TF_NewStatus();
+        TF_SessionOptions* SessionOpts = TF_NewSessionOptions();
+        TF_Buffer* RunOpts = NULL;
+        
+        tags = "serve"; 
+        
+        int ntags = 1; 
+        Session = TF_LoadSessionFromSavedModel(SessionOpts, RunOpts, saved_model_dir, &tags, ntags, Graph, NULL, Status);
+        
+        if (TF_GetCode(Status) == TF_OK) {
+            std::cout << "TF_LoadSessionFromSavedModel OK\n" << std::endl;
+        }
+        else {
+            std::cout << "%s" << TF_Message(Status) << std::endl;
+        }
+    
+
+        //********* Get input tensor
+        Input = (TF_Output*) malloc(sizeof(TF_Output) * NumInputs);
+        t0 = {TF_GraphOperationByName(Graph, "serving_default_input"), 0};
+
+        if(t0.oper == NULL) {
+            std::cout << "ERROR: Failed TF_GraphOperationByName serving_default_input\n" << std::endl;
+        }
+        else {
+            std::cout << "TF_GraphOperationByName serving_default_input is OK\n" << std::endl;
+        }
+        Input[0] = t0;
+
+        
+        //********* Get Output tensor
+        Output = (TF_Output*) malloc(sizeof(TF_Output) * NumOutputs);
+        t2 = {TF_GraphOperationByName(Graph, "StatefulPartitionedCall"), 0};
+        
+        if(t2.oper == NULL) {
+            std::cout << "ERROR: Failed TF_GraphOperationByName StatefulPartitionedCall\n" << std::endl;
+        }
+        else{
+            std::cout << "TF_GraphOperationByName StatefulPartitionedCall is OK\n" << std::endl;
+        }
+        Output[0] = t2; 
+
+
+        //********* Allocate data for inputs & outputs
+        InputValues  = (TF_Tensor**)malloc(sizeof(TF_Tensor*)*NumInputs);
+        OutputValues = (TF_Tensor**)malloc(sizeof(TF_Tensor*)*NumOutputs);
+
+    }
+
+std::vector<float> ModelLoader::Evaluator(std::vector<float> neural_network_input, int num_actions)
+    {        
+        int ndims = 2; 
```

Comment:
what is ndims, why no parameter, what is this?

Reply:
It’s the number of dimension of the neural_network_input vector